Reorganize subsystems/commands to be more readable & consistent

5t0n3 opened the pull request:
For commands, I tried to rename them to better reflect what they do. I've also started trying to migrate away from `InstantCommand`s since they're meant to be used in only one place, and we were repeating them. Commands were also moved to the respective folders based on the subsystems they act on, or the utilities folder if they don't act on a subsystem.

In the way of subsystems:
* I removed some unused code from `ShooterSubsystem`, and moved calculations to `utilities/ShooterMath.java` (although this file only has one method).
* I tried to document the methods in `DriveSubsystem` more, as well as added odometry back in for use with RAMSETE control.

I haven't really touched the other subsystems yet, but I might in future commits.

We should also definitely test these changes before merging them into `main`.

Commit:
Add an auto that taxis and shoots a preloaded ball

Commit:
Turn limelight on at robot init and never turn it off

Commit:
Add ClimbEasyHighBar command (courtesy of Nolan)

Commit:
Use ClimbEasyHighBar in telop

Commit:
Changes to TaxiThenShoot and ClimbEasyHighBar

TaxiThenShoot (and by extension RampThenShoot) no longer depend on supplying a
controller.

ClimbEasyHighBar and ClimbTimed timings have been tuned a bit.

Timeouts were also added for AlignWithHubSmart because it wasn't
reaching its goal, preventing drivers from driving the robot.

Commit:
Make TaxiThenShoot the default auto choice

Also added in some delays to ClimbTimed.

The default angle in ShooterSubsystem was changed to 12 from 5 as well.

Commit:
Lots of different changes

Increased feedforward constant for turning with Smart Motion because of
increase friction due to the carpet.

Increased the current limit for the intake arm to 15 amps, and Increased
its max power to full power.

Undo port forwarding because it might be causing robot code to cut out?

Added reversal for intake and indexer.

Intake now runs for all of TaxiThenShoot autonomous.

Also fixed RampThenShoot, as it was throwing a NullPointerException in
the past.

Intake lift motors were changed to coast mode from brake mode to avoid
shredding the gears (they're aluminum).

Commit:
Further tuned climb timings and turn feedforward

Commit:
Extend TaxiThenShoot distance

Commit:
Again, a bunch of different competition changes

Added arbitrary feedforward power for turning using Smart Motion.

Secondary camera feed is now functional, but for some reason the
limelight feed doesn't work on the driver station.

Adjusted ClimbTimed timings again.

Started implementing speed switching in DefaultDrive but didn't finish.

IntakeLift is meant to be used when the intake lift motors are in brake
mode to reduce wear on the gears.

Shooter ramp time in RampThenShoot was increased to a maximum of 1
second.

Commit:
Merge branch 'main' into wilsonville

Commit:
Make Limelight wrapper not a subsystem, since its use isn't exclusive

Commit:
Uncommitted refactoring changes

Commit:
Reorganize & remove unnecessary shooter code

Also move calculations to utilities/ShooterMath.java, although that
might not be necessary.

Commit:
Move commands to their respective folders

Commit:
Formatting!

Commit:
Separate DefaultDrive and SwappableAccelDrive

Commit:
Make Limelight control commands actually end

Commit:
Delete HomeOnHub since it is no longer used

Commit:
Add odometry back to DriveSubsystem & document its methods

Commit:
Rename TeleOpShooter to RunShooterWithLimelight

Commit:
Delete LimeLightTargeting, as it's been superseded by the Limelight class

Commit:
Merge branch 'main' into reorganization

I don't know why this was necessary but here we are.

nolanhergert commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -143,13 +143,10 @@ private void configureButtonBindings() {
-  public void turnOnLimelightLED() {
-    m_limelight.turnOnLED();
+        .whileHeld(
+            new ParallelCommandGroup(
```
I know you didn't change this, but I wouldn't run the upper indexer here. It should only run the lower for now, until we get a working color sensor / detection so we don't run the ball too high.

5t0n3 replied:
I think the purpose of this was to eject the ball, not run it into the shooter, which was why we ran both the upper and lower indexer.

nolanhergert commented on src/main/java/frc/robot/commands/autonomous/TaxiThenShoot.java:
```diff
@@ -6,30 +6,35 @@
+      Limelight limelight) {
+    addCommands(
+        // Turn on the intake
+        new InstantCommand(() -> intake.runRollerAtMaxPower(false), intake),
```
Unrelated to this, but I would rename the functions to runRoller(), runRollerBackwards(), and stopRoller(). We shouldn't set intake roller RPM directly.

nolanhergert commented on src/main/java/frc/robot/commands/climber/ClimbEasyHighBar.java:
```diff
@@ -2,29 +2,25 @@
 // the WPILib BSD license file in the root directory of this project.
 
-package frc.robot.commands;
+package frc.robot.commands.climber;
```
You can delete this entire file

nolanhergert commented on src/main/java/frc/robot/commands/drive/AlignWithHub.java:
```diff
@@ -44,7 +45,7 @@ public AlignWithHub(DriveSubsystem drive, LimeLightTargeting limelight) {
     // Make sure to reset the heading before resetting the internal PID controller
     m_drive.resetHeading();
-    m_limelight.turnOnLED();
+    m_limelight.setLEDMode(LEDMode.On);
```
I like the naming of turnOnLED() better than setLEDMode().  But I see there's also a  TurnOnLimelightLEDs()?

5t0n3 replied:
`TurnOnLImelightLEDs` was the command form of turning on the Limelight LEDs. The reason I named it `setLEDMode` was to reflect that you pass it an enum instead of just a number (I think I've brought that up in the past). Given that we only ever use off and on (not pipeline & blink) the enum might be overkill to use outside of the `Limelight` class, but I'll probably leave them in just in case.

nolanhergert commented on src/main/java/frc/robot/commands/drive/DefaultDrive.java:
```diff
@@ -16,75 +12,31 @@
-  // TODO: Move this to DriveSubsystem (using the setDeadband method)
-  private double deadzone = 0.1;
+  // TODO: Move to Constants.java?
+  private double m_speedMultiplier = 1.0;
```
m_forwardBackwardMultiplier? 

Also I think it's ok to move to Constants.java too.

nolanhergert commented on src/main/java/frc/robot/commands/drive/DefaultDrive.java:
```diff
@@ -16,75 +12,31 @@
-      defenseMode = !defenseMode;
-    }
+    // Calculate appropriate powers using above slew limiters
+    double forwardPower = m_speedMultiplier * -m_left.getAsDouble();
```
We still need deadband, and you can technically combine 34 and 35 too, it seems to work fine.

nolanhergert commented on src/main/java/frc/robot/commands/drive/DefaultDrive.java:
```diff
@@ -16,75 +12,31 @@
-  SlewRateLimiter defenseSlewLimiter = new SlewRateLimiter(2.5);
+  // Acceleration limiter for forward/rotational movement
+  SlewRateLimiter m_forwardSlewLimiter = new SlewRateLimiter(2.0);
+  SlewRateLimiter m_turningSlewLimiter = new SlewRateLimiter(2.0);
```
Please change name to rotationSlewLimiter to keep it consistent. I probably made it that way initially.

nolanhergert commented on src/main/java/frc/robot/commands/drive/SwappableAccelDrive.java:
```diff
@@ -0,0 +1,71 @@
+package frc.robot.commands.drive;
```
I think any changes you made here should be part of DefaultDrive and not a near copy paste.

5t0n3 replied:
I'll probably test the slew swapping on the bot first, but after verifying that it works I'll just move it to DefaultDrive.

nolanhergert commented on src/main/java/frc/robot/commands/indexer/RunUpperIndexer.java:
```diff
@@ -4,26 +4,26 @@
-    }
+  @Override
+  public void execute() {
+    m_indexer.powerExitSide(powerMultiplier * 0.7);
```
Not your problem, but I think power/rpm should be set in the subsystem for everything except shooter to keep things simplified.

nolanhergert commented on src/main/java/frc/robot/commands/intake/LowerAndSpinIntake.java:
```diff
@@ -4,23 +4,23 @@
 import frc.robot.subsystems.IntakeSubsystem;
```
There are no references to this file, let's delete it to keep things simple.

nolanhergert commented on src/main/java/frc/robot/commands/utilities/TurnOffLimelightLEDs.java:
```diff
@@ -0,0 +1,23 @@
+package frc.robot.commands.utilities;
```
I don't get the point of turn off / turn on LEDs files. IMO the one line that does anything should be a part of homing/shooting and that's it. Don't need a separate command, I think it's fine to be an instant command, it shouldn't be managed by the drive team, etc.

nolanhergert commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -87,11 +73,11 @@ public DriveSubsystem() {
-    // Initialize the tracking of the robot's position on the field
-    // m_odometry = new DifferentialDriveOdometry(new Rotation2d(getHeading()));
+    // Configure a deadband of 0.1 for arcade/tank drive
+    m_drive.setDeadband(0.1);
```
We need a deadband on the XBoxController side, not for any other uses of the Drive Subsystem.

5t0n3 replied:
I guess the variable naming is a bit ambiguous here, but in this case the deadband is being applied to the `DifferentialDrive` instance within `DriveSubsystem`, which is responsible for arcade/tank drive using the sticks. Since that's the only place where the sticks are used, I figured we could move it there instead of deadbanding them individually.

nolanhergert commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -272,8 +241,10 @@ public void smartMotionToPosition(double left, double right, boolean turning) {
-      m_rightController.setReference(right, ControlType.kSmartMotion, pidSlot, 0.04, ArbFFUnits.kPercentOut);
-      m_leftController.setReference(left, ControlType.kSmartMotion, pidSlot, 0.04, ArbFFUnits.kPercentOut);
+      m_rightController.setReference(
+          right, ControlType.kSmartMotion, pidSlot, 0.04, ArbFFUnits.kPercentOut);
```
Ahh, this .04 is where the magic static feed forward offset is inputted. We should definitely make a variable in Constants out of this and use it especially for turning.

5t0n3 replied:
0.04 also wasn't a value that worked, although we should still have the old value somewhere. I'll make sure to test that tomorrow :)

Commit:
Delete a couple unused commands

Commit:
Remove Turn{Off,On}LimelightLEDs commands

Commit:
Rewrite IntakeSubsystem to be more ergonomic

This also required a few changes elsewhere due to removed/renamed
methods.

Commit:
Intake subsystem rewrite

This included the renaming of some methods & changes elsewhere to
reflect these new names.

An enum switch is now used for RunUpperIndexer, just as with
RunIntakeRoller.

Commit:
Move more drive-related constants to Constants.java

Commit:
Got Limelight-based alignment with hub functional

Also fixed a few NullPointerExceptions and the like elsewhere.

Commit:
Move joystick deadbanding back into driving commands
